### PR TITLE
NETOBSERV-2005: Support multiple flow filter rules using json fmt string

### PR DESCRIPTION
**Pull request by msherif1234**
## Description

Example config and the flow filter map content
```sh
oc get pods  -n netobserv-privileged netobserv-ebpf-agent-tksvd -o yaml | grep "FLOW_FILTER_RULES" -A1
    - name: FLOW_FILTER_RULES
      value: '[{"ip_cidr":"1.1.1.1/24","protocol":"TCP","source_port":53,"action":"Accept","tcp_flags":"SYN"},{"ip_cidr":"2.2.2.2/24","protocol":"UDP","source_port":80,"action":"Accept","tcp_flags":"SYN"},{"ip_cidr":"3.3.3.3/24","protocol":"ICMP","action":"Accept"}]'

bpftool map dump id 223 
[{
        "key": {
            "prefix_len": 24,
            "ip_data": [1,1,1,1,0,0,0,0,0,0,0,0,0,0,0,0
            ]
        },
        "value": {
            "protocol": 6,
            "dstPortStart": 0,
            "dstPortEnd": 0,
            "dstPort1": 0,
            "dstPort2": 0,
            "srcPortStart": 53,
            "srcPortEnd": 0,
            "srcPort1": 0,
            "srcPort2": 0,
            "portStart": 0,
            "portEnd": 0,
            "port1": 0,
            "port2": 0,
            "icmpType": 0,
            "icmpCode": 0,
            "direction": "MAX_DIRECTION",
            "action": "ACCEPT",
            "tcpFlags": "SYN_FLAG",
            "filter_drops": 0,
            "ip": [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0
            ]
        }
    },{
        "key": {
            "prefix_len": 24,
            "ip_data": [2,2,2,2,0,0,0,0,0,0,0,0,0,0,0,0
            ]
        },
        "value": {
            "protocol": 17,
            "dstPortStart": 0,
            "dstPortEnd": 0,
            "dstPort1": 0,
            "dstPort2": 0,
            "srcPortStart": 80,
            "srcPortEnd": 0,
            "srcPort1": 0,
            "srcPort2": 0,
            "portStart": 0,
            "portEnd": 0,
            "port1": 0,
            "port2": 0,
            "icmpType": 0,
            "icmpCode": 0,
            "direction": "MAX_DIRECTION",
            "action": "ACCEPT",
            "tcpFlags": "SYN_FLAG",
            "filter_drops": 0,
            "ip": [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0
            ]
        }
    },{
        "key": {
            "prefix_len": 24,
            "ip_data": [3,3,3,3,0,0,0,0,0,0,0,0,0,0,0,0
            ]
        },
        "value": {
            "protocol": 1,
            "dstPortStart": 0,
            "dstPortEnd": 0,
            "dstPort1": 0,
            "dstPort2": 0,
            "srcPortStart": 0,
            "srcPortEnd": 0,
            "srcPort1": 0,
            "srcPort2": 0,
            "portStart": 0,
            "portEnd": 0,
            "port1": 0,
            "port2": 0,
            "icmpType": 0,
            "icmpCode": 0,
            "direction": "MAX_DIRECTION",
            "action": "ACCEPT",
            "tcpFlags": 0,
            "filter_drops": 0,
            "ip": [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0
            ]
        }
    }
]

 ```
## Dependencies

n/a

## Checklist

If you are not familiar with our processes or don't know what to answer in the list below, let us know in a comment: the maintainers will take care of that.

* [ ] Will this change affect NetObserv / Network Observability operator? If not, you can ignore the rest of this checklist.
* [ ] Is this PR backed with a JIRA ticket? If so, make sure it is written as a title prefix _(in general, PRs affecting the NetObserv/Network Observability product should be backed with a JIRA ticket - especially if they bring user facing changes)._
* [ ] Does this PR require product documentation?
  * [ ] If so, make sure the JIRA epic is labelled with "documentation" and provides a description relevant for doc writers, such as use cases or scenarios. Any required step to activate or configure the feature should be documented there, such as new CRD knobs.
* [ ] Does this PR require a product release notes entry?
  * [ ] If so, fill in "Release Note Text" in the JIRA.
* [ ] Is there anything else the QE team should know before testing? E.g: configuration changes, environment setup, etc.
  * [ ] If so, make sure it is described in the JIRA ticket.
* QE requirements (check 1 from the list):
  * [ ] Standard QE validation, with pre-merge tests unless stated otherwise.
  * [ ] Regression tests only (e.g. refactoring with no user-facing change).
  * [ ] No QE (e.g. trivial change with high reviewer's confidence, or per agreement with the QE team).

### CODE REVIEW

File: pkg/agent/agent.go [hunk 1]
```diff
@@ -199,7 +200,28 @@ func FlowsAgent(cfg *Config) (*Flows, error) {
...
+			filterRules = append(filterRules, &tracer.FilterConfig{
+				FilterAction:          r.FilterAction,
+				FilterDirection:       r.FilterDirection,
+				FilterIPCIDR:          r.FilterIPCIDR,
+				FilterProtocol:        r.FilterProtocol,
+				FilterPeerIP:          r.FilterPeerIP,
+				FilterDestinationPort: tracer.ConvertFilterPortsToInstr(r.FilterDestinationPort, r.FilterDestinationPortRange, r.FilterDestinationPorts),
+				FilterSourcePort:      tracer.ConvertFilterPortsToInstr(r.FilterSourcePort, r.FilterSourcePortRange, r.FilterSourcePorts),
+				FilterPort:            tracer.ConvertFilterPortsToInstr(r.FilterPort, r.FilterPortRange, r.FilterPorts),
+				FilterTCPFLags:        r.FilterTCPFlags,
```
Comment (stleerh) [inline pkg/agent/agent.go:220]: FilterTCPFlags (lowercase 'l' in "Flags").  Do a find/replace.

File: pkg/agent/config.go [hunk 2]
```diff
@@ -26,6 +26,57 @@ const (
...
 
+type FlowFilter struct {
+	// FilterDirection is the direction of the flow filter.
+	// Possible values are "Ingress" or "Egress".
+	FilterDirection string `json:"direction,omitempty"`
+	// FilterIPCIDR is the IP CIDR to filter flows.
+	// Example: 10.10.10.0/24 or 100:100:100:100::/64, default is 0.0.0.0/0
+	FilterIPCIDR string `json:"ip_cidr,omitempty"`
+	// FilterProtocol is the protocol to filter flows.
+	// Example: TCP, UDP, SCTP, ICMP, ICMPv6
```
Comment (stleerh) [inline pkg/agent/config.go:37]: If these are the only possible values, state that instead of as examples.

**Review by stleerh — COMMENTED** [review]
Minor comments.
/lgtm

**Review by jpinsonneau — APPROVED** [review]
LGTM, thanks @msherif1234 !

File: bpf/flows.c [hunk 3]
```diff
@@ -77,15 +79,7 @@ static inline void update_dns(additional_metrics *extra_metrics, pkt_info *pkt,
 }
 
 static inline int flow_monitor(struct __sk_buff *skb, u8 direction) {
-    // If sampling is defined, will only parse 1 out of "sampling" flows
```
Comment (jotak) [inline bpf/flows.c:80]: we could have optimized this, and if no filter_sampling is defined, run the global sampling check earlier. Keep the late sampling check only when there filter-based sampling.

Comment (jotak) [reply bpf/flows.c:80]: Running some perf, while user-space metrics look unchanged, using bpftop I do see a consequent increase in CPU in kernel, around +100%, when using global sampling=50 and no filter sampling, compared to before main baseline this PR was merged.

**Pull request merged**